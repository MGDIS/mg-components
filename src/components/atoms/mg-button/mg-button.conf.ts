/**
 * List of all possibles variants
 */
export const variants = ['primary', 'secondary', 'danger', 'danger-alt', 'info', 'flat', 'success', 'link'] as const;

/**
 * VariantType type from button variants
 */
export type VariantType = typeof variants[number];

/**
 * List of all possibles button types
 */
export const buttonTypes = ['button', 'submit', 'reset'] as const;

/**
 * ButtonType type from buttonTypes
 */
export type ButtonType = typeof buttonTypes[number];
