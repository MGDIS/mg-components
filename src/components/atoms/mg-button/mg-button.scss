/* Variables */
$mg-button-border-width: 0.1rem;
$mg-button-padding-horizontal: 1.1rem;
$button-column-gap: 0.6rem;

/* Mixins */
@mixin style-gradient($name, $backgroundColor, $fontColor, $hasVariable: false) {
  @if $hasVariable {
    $backgroundColor: --mg-button-#{$name}-color;
    $fontColor: --mg-button-#{$name}-font-color;
  } @else {
    $backgroundColor: --color-#{$backgroundColor};
    $fontColor: --color-#{$fontColor};
  }
  .mg-button.mg-button--#{$name} {
    // Decoration
    border-color: hsl(
      var(
        #{--mg-button-#{$name}-border-color},
        (var(#{$backgroundColor}-h), var(#{$backgroundColor}-s), calc(var(#{$backgroundColor}-l) - var(--mg-button-border-variation) * 10%))
      )
    );
    // Font
    color: hsl(var(#{$fontColor}));
    // Background
    background: linear-gradient(
      to bottom,
      hsl(var(#{$backgroundColor}-h), var(#{$backgroundColor}-s), var(#{$backgroundColor}-l)),
      hsl(var(#{$backgroundColor}-h), var(#{$backgroundColor}-s), calc(var(#{$backgroundColor}-l) - var(--mg-button-gradient) * 10%))
    );
  }
  // Only apply if not disabled
  :host(:hover),
  :host(:focus),
  :host(:active) {
    .mg-button--#{$name}:not(.mg-button--disabled) {
      background: hsl(var(#{$backgroundColor}-h), var(#{$backgroundColor}-s), var(#{$backgroundColor}-l));
    }
  }
}

@mixin font-style() {
  text-decoration: none;
  font-family: inherit;
  font-weight: var(--mg-button-font-weight, normal);
  text-align: left;
  font-size: var(--font-size);
}

:host {
  display: inline-block;
  vertical-align: middle;
  max-width: 100%;
}

:host(:focus-visible) {
  outline: 0;
  // same style as chrome
  .mg-button:not(.mg-button--disabled) {
    outline-color: -webkit-focus-ring-color;
    outline-style: auto;
    outline-width: 0.2rem;
  }
}

// Only apply if not disabled
:host(:hover),
:host(:focus),
:host(:active) {
  .mg-button--flat:not(.mg-button--disabled) {
    background: hsla(var(--color-dark), 0.1);
  }
}

:host([full-width]) {
  display: block;
}

// Gradient style
@include style-gradient('primary', null, null, true);
@include style-gradient('secondary', null, null, true);
@include style-gradient('danger', 'danger', 'neutral');
@include style-gradient('danger-alt', 'neutral', 'danger');
@include style-gradient('info', 'info', 'neutral');
@include style-gradient('success', 'success', 'dark');

/* Styles */
.mg-button {
  // Display
  position: relative;
  display: inline-flex;
  align-items: center;
  column-gap: $button-column-gap;
  padding: calc((var(--default-size) - (2 * $mg-button-border-width) - (var(--font-size) * var(--line-height))) / 2) $mg-button-padding-horizontal;
  min-height: var(--default-size);
  box-sizing: border-box;
  // Decoration
  border-top-right-radius: var(--mg-button-border-radius-top-right, var(--mg-button-border-radius));
  border-top-left-radius: var(--mg-button-border-radius-top-left, var(--mg-button-border-radius));
  border-bottom-right-radius: var(--mg-button-border-radius-bottom-right, var(--mg-button-border-radius));
  border-bottom-left-radius: var(--mg-button-border-radius-bottom-left, var(--mg-button-border-radius));
  cursor: pointer;
  background-image: none;
  border-top-width: var(--mg-button-border-top-width, $mg-button-border-width);
  border-right-width: var(--mg-button-border-right-width, $mg-button-border-width);
  border-bottom-width: var(--mg-button-border-bottom-width, $mg-button-border-width);
  border-left-width: var(--mg-button-border-left-width, $mg-button-border-width);
  border-style: solid;
  // Font (necessary for stylesheet only)
  @include font-style();
  // Other
  touch-action: manipulation;

  // Disabled style
  &.mg-button--disabled {
    opacity: var(--mg-button-disabled-opacity);
    cursor: default;
  }

  // Icon only button
  &.mg-button--icon {
    // Display
    padding: 0;
    justify-content: center;
    width: var(--default-size);
    height: var(--default-size);
    // Decoration
    border-top-right-radius: var(--mg-button-border-radius-top-right, var(--mg-button-icon-border-radius));
    border-top-left-radius: var(--mg-button-border-radius-top-left, var(--mg-button-icon-border-radius));
    border-bottom-right-radius: var(--mg-button-border-radius-bottom-right, var(--mg-button-icon-border-radius));
    border-bottom-left-radius: var(--mg-button-border-radius-bottom-left, var(--mg-button-icon-border-radius));
    // font
    font-size: var(--font-size);
    line-height: var(--font-size);
  }

  // Flat style
  &.mg-button--flat {
    background: none;
    border: none;
  }
  // full-width style
  &.mg-button--full-width {
    display: flex;
    justify-content: center;
  }

  // loading style
  &.mg-button--loading {
    > mg-icon:first-of-type {
      position: absolute;
      left: 50%;
      top: 50%;
      transform: translate(-50%, -50%);
    }
    .mg-button__content {
      opacity: 0;
    }
  }

  mg-icon:only-child,
  ::slotted(mg-icon) {
    display: inline-flex;
    flex-shrink: 0;
    line-height: 1;
    margin: auto 0;
  }

  slot,
  ::slotted(*:not(mg-icon)) {
    // Font
    @include font-style();
    // Display
    display: flex;
    align-items: center;
    column-gap: $button-column-gap;
  }
}
