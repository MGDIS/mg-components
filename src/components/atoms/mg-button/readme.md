This atom refers to the [PDA9-67](https://jira.mgdis.fr/browse/PDA9-67).

<!-- Auto Generated Below -->


## Properties

| Property         | Attribute          | Description                                                                                                                                            | Type      | Default                 |
| ---------------- | ------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------ | --------- | ----------------------- |
| `disableOnClick` | `disable-on-click` | Option to set input disable on click, in order to prevent multi-click. Parent component have to remove the attribute 'disabled' when the process ends. | `boolean` | `false`                 |
| `disabled`       | `disabled`         | Disable button                                                                                                                                         | `boolean` | `false`                 |
| `identifier`     | `identifier`       | Identifier is used for the element ID (id is a reserved prop in Stencil.js) If not set, it will be created.                                            | `string`  | `createID('mg-button')` |
| `isIcon`         | `is-icon`          | Define if button is round. Used for icon button.                                                                                                       | `boolean` | `false`                 |
| `label`          | `label`            | aria-label In case button text is not explicit enough                                                                                                  | `string`  | `undefined`             |
| `variant`        | `variant`          | Define button variant                                                                                                                                  | `string`  | `variants[0]`           |


## Dependencies

### Used by

 - [mg-message](../../molecules/mg-message)
 - [mg-pagination](../../molecules/mg-pagination)

### Depends on

- [mg-icon](../mg-icon)

### Graph
```mermaid
graph TD;
  mg-button --> mg-icon
  mg-message --> mg-button
  mg-pagination --> mg-button
  style mg-button fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
