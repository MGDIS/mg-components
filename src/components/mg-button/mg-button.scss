@mixin button-state($variant, $state: '') {
  @if $state != '' {
    $state: -#{$state};
  }
  // Decoration
  border-color: var(--button-#{$variant}#{$state}-border-color);
  // Font
  color: var(--button-#{$variant}#{$state}-color);
}

@mixin button-variant($variant) {
  @include button-state($variant);
  background: linear-gradient(to bottom, var(--button-#{$variant}-background-from), var(--button-#{$variant}-background-to));

  // Hover
  &:hover,
  &:focus,
  &:active {
    @include button-state($variant, 'hover');
    background: var(--button-#{$variant}-hover-background);
  }

  // Disabled
  &[disabled],
  &.disabled {
    @include button-state($variant, 'disabled');
    background: linear-gradient(to bottom, var(--button-#{$variant}-disabled-background-from), var(--button-#{$variant}-disabled-background-to));
    opacity: var(--button-#{$variant}-disabled-opacity);
  }
}

@mixin button-variants($variants...) {
  @each $variant in $variants {
    &.mg-button--#{$variant} {
      @include button-variant($variant);
    }
  }
}

.mg-button {
  // Display
  display: inline-block;
  min-height: 3.5rem;
  padding: 0.6rem 1.2rem;
  vertical-align: middle;
  // Decoration
  border-radius: 0.3rem;
  cursor: pointer;
  background-image: none;
  border: 0.1rem solid transparent;
  // Font
  font-weight: normal;
  text-align: center;
  white-space: nowrap;
  // Other
  touch-action: manipulation;

  &[disabled],
  &.disabled {
    cursor: not-allowed;
  }

  @include button-variants('primary', 'secondary', 'alert', 'alert-alt', 'info');
}

mg-button {
  mg-icon {
    margin-right: 0.5rem;
  }
  & + mg-button {
    margin-left: 1rem;
  }
}
