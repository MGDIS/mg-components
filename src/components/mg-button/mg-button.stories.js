export default {
  title: 'Demo/MgButton',
  argTypes: {
    label: { type: 'text', description: 'The text which is shown as label' },
    name: {
      type: 'text',
      description: 'Is needed to reference the form data after the form is submitted',
    },
    disabled: {
      type: 'boolean',
      description: 'If true, the button is displayed as disabled',
      defaultValue: { summary: false },
    },
  },
};

const defaultArgs = {
  disabled: false,
};

const Template = args => {
  return <mg-button {...args}></mg-button>;
};

export const MgButton = Template.bind({});
Default.MgButton = { ...defaultArgs };
