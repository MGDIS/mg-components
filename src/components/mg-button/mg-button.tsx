import { Component, Host, h, Prop, Watch } from '@stencil/core';

@Component({
  tag: 'mg-button',
  styleUrl: 'mg-button.scss',
  shadow: true,
})
export class MgButton {

  /**
   * Internal
   */
  private classes = ['mg-button'];
  private variants = ["primary" , "secondary" , "alert", "alert-alt",  "info"]; // TODO : should we have a link styled button ?

  /**
   * Define button style
   */
  @Prop() variant: string = "primary";
  @Watch('variant')
  validateVariant(newValue: string) {
    if(!this.variants.includes(newValue)) {
      throw new Error(`<mg-button> prop "variant" must be one of : ${this.variants.join(', ')}`);
    }
    this.classes.push(`mg-button-${this.variant}`);
  }

  /**
   * aria-label
   * In case button text is not explicit enougth
   */
   @Prop() label: string;

  /**
   * Disable button
   */
   @Prop() disabled: boolean = false;

  /**
   * Check if props are well configured on init
   */
  componentWillLoad() {
    this.validateVariant(this.variant);
  }

  /**
  * Render component
  */
  render() {
    return (
      <Host>
        <button class={this.classes.join(' ')} aria-label={this.label} disabled={this.disabled}><slot></slot></button>
      </Host>
    );
  }

}
