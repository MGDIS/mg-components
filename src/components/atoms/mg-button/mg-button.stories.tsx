import { h } from '@stencil/core';
import { variants } from './mg-button.conf';
import { icons } from '../mg-icon/mg-icon.conf';

export default {
  component: 'mg-button',
  title: 'Atoms/mg-button',
  argTypes: {
    variant: {
      options: variants,
      control: { type: 'select' },
    },
  },
};

const Template = args => {
  // Extract slot so it won't be render as an attribute
  const slot = args.slot;
  delete args.slot;
  const disableOnClick = args.disableOnClick;
  delete args.disableOnClick;
  const isIcon = args.isIcon;
  delete args.isIcon;
  // return element
  return (
    <mg-button {...args} disable-on-click={disableOnClick} is-icon={isIcon}>
      {slot}
    </mg-button>
  );
};

export const MgButton = Template.bind({});

MgButton.args = {
  slot: 'Text button',
  variant: variants[0],
  label: 'Explicit aria label',
  identifier: undefined,
  disabled: false,
  disableOnClick: false,
  isIcon: false,
};

export const IsIcon = Template.bind({});

IsIcon.args = {
  ...MgButton.args,
  isIcon: true,
  slot: <mg-icon icon={Object.keys(icons)[0]}></mg-icon>,
};

export const WithBadge = Template.bind({});

WithBadge.args = {
  ...MgButton.args,
  slot: (
    <span>
      Button with badge <mg-badge>23</mg-badge>
    </span>
  ),
};

export const DisableOnClick = Template.bind({});

DisableOnClick.args = {
  ...MgButton.args,
  disableOnClick: true,
};
