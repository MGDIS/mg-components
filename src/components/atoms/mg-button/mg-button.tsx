import { Component, Element, h, Prop, State, Watch, Host } from '@stencil/core';
import { variants, VariantType, ButtonType } from './mg-button.conf';
import { ClassList } from '../../../utils/components.utils';

@Component({
  tag: 'mg-button',
  styleUrl: 'mg-button.scss',
  shadow: true,
})
export class MgButton {
  /************
   * Internal *
   ************/

  private onClickElementFn = null;
  private classDisabled = 'mg-button--disabled';
  private classLoading = 'mg-button--loading';
  private classFullWidth = 'mg-button--full-width';

  /**************
   * Decorators *
   **************/

  /**
   * Get component DOM element
   */
  @Element() element: HTMLMgButtonElement;

  /**
   * Define button variant
   */
  @Prop() variant: VariantType = variants[0]; // Primary
  @Watch('variant')
  validateVariant(newValue: VariantType, oldValue?: VariantType): void {
    if (!variants.includes(newValue)) {
      throw new Error(`<mg-button> prop "variant" must be one of : ${variants.join(', ')}`);
    } else {
      if (oldValue !== undefined) {
        this.classList.delete(`mg-button--${oldValue}`);
      }
      this.classList.add(`mg-button--${newValue}`);
    }
  }

  /**
   * Identifier is used for the element ID (id is a reserved prop in Stencil.js)
   */
  @Prop() identifier: string;

  /**
   * aria-label
   * In case button text is not explicit enough
   */
  @Prop() label: string;

  /**
   * Define button type
   */
  @Prop() type: ButtonType;

  /**
   * Set button to full-width
   */
  @Prop({ mutable: true }) fullWidth = false;
  @Watch('fullWidth')
  validateFullWidth(newValue) {
    if (newValue && this.isIcon) {
      this.classList.delete(this.classFullWidth);
      throw new Error('<mg-button> prop "fullWidth" cannot be used with prop "isIcon".');
    } else if (newValue) {
      this.classList.add(this.classFullWidth);
    } else {
      this.classList.delete(this.classFullWidth);
    }
  }

  /**
   * Define form id to attach button with.
   * If this attribute is not set, the <button> is associated with its ancestor <form> element.
   */
  @Prop({ mutable: true }) form: string;

  /**
   * Disable button
   */
  @Prop({ mutable: true }) disabled: boolean;
  @Watch('disabled')
  disabledHandler(isDisabled: boolean): void {
    // Remove loading when enable
    // Will be set back onclick
    if (!isDisabled && this.disableOnClick) {
      this.loading = false;
    }
    // Manage if onclick
    this.element.onclick = isDisabled ? undefined : this.onClickElementFn;

    // apply style
    if (isDisabled) {
      this.classList.add(this.classDisabled);
    } else {
      this.classList.delete(this.classDisabled);
    }
  }
  /**
   * Define if button is round.
   * Used for icon button.
   */
  @Prop() isIcon = false;

  /**
   * Option to set input disable on click, in order to prevent multi-click.
   * Parent component have to remove the attribute 'disabled' when the process ends.
   */
  @Prop() disableOnClick = false;

  /**
   * Define if button is loading, default to false.
   * Trigger when button is clicked or key-up ['enter', 'space], then value change to true.
   * It's required to reset to false when action/promise in parent is done to stop the loading state
   */
  @State() loading = false;
  @Watch('loading')
  loadingHandler(newValue: boolean): void {
    // we add loading style if it newvalue is true else we remove it
    if (newValue) {
      this.classList.add(this.classLoading);
    } else {
      this.classList.delete(this.classLoading);
    }
  }

  /**
   * Component classes
   */
  @State() classList: ClassList = new ClassList(['mg-button']);

  /**
   * Trigger actions onClick event
   *
   * @param {MouseEvent} event click event
   * @returns {void}
   */
  private handleClick = (event: MouseEvent): void => {
    if (this.disabled) event.stopPropagation();
    // Used to prevent multi-click.
    else if (this.disableOnClick) {
      this.loading = true;
      this.disabled = true;
    }
  };

  /**
   * Handle onKeyDown event
   *
   * @param {KeyboardEvent} event keyboard event
   * @returns {void}
   */
  private handleKeydown = (event: KeyboardEvent): void => {
    if (!this.disabled && event.key === ' ') {
      event.preventDefault();
    } else if (!this.disabled && ['Enter', 'NumpadEnter', 'Space'].includes(event.key)) {
      event.preventDefault();
      this.element.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    }
  };

  /**
   * Handle onKeyUp event
   *
   * @param {KeyboardEvent} event keyboard event
   * @returns {void}
   */
  private handleKeyup = (event: KeyboardEvent): void => {
    if (!this.disabled && event.key === ' ') {
      event.preventDefault();
      this.element.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    }
  };

  /**
   * Check if props are well configured on init
   *
   * @returns {void}
   */
  componentWillLoad(): void {
    this.validateVariant(this.variant);
    this.validateFullWidth(this.fullWidth);
    if (this.isIcon) {
      this.classList.add(`mg-button--icon`);
      if (typeof this.label !== 'string' || this.label.trim() === '') {
        throw new Error(`<mg-button> prop "label" is mandatory when prop "isIcon" is set to true.`);
      }
    }
    // Store the onclick fn
    this.onClickElementFn = this.element.onclick;
    this.disabledHandler(this.disabled);
  }

  /**
   * Render component
   *
   * @returns {HTMLElement} html element
   */
  render(): HTMLElement {
    return (
      <Host
        role="button"
        tabIndex={this.disabled ? -1 : 0}
        type={this.type}
        form={this.form}
        mgWidth={this.fullWidth === true && 'full'} // css selector for host to prevent class usage bug on it
        aria-label={this.label}
        aria-disabled={this.disabled !== undefined && this.disabled.toString()}
        onClick={this.handleClick}
        onKeyUp={this.handleKeyup}
        onKeyDown={this.handleKeydown}
      >
        <div class={this.classList.join()} id={this.identifier}>
          {this.loading && <mg-icon icon="loader" spin></mg-icon>}
          <div class="mg-button__content">
            <slot></slot>
          </div>
        </div>
      </Host>
    );
  }
}
