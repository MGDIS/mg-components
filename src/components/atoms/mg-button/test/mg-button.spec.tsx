import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgButton } from '../mg-button';
import { variants, buttonTypes } from '../mg-button.conf';

const getPage = (args, content = 'Text button') =>
  newSpecPage({
    components: [MgButton],
    template: () => <mg-button {...args}>{content}</mg-button>,
  });

describe('mg-button', () => {
  test('Should render a button with an id', async () => {
    const { root } = await getPage({ identifier: 'identifier', label: 'label' });
    expect(root).toMatchSnapshot();
  });

  describe.each(variants)('Should render an %s button', variant => {
    test.each([false, true])('isIcon %s', async isIcon => {
      const { root } = await getPage({ variant, isIcon, label: 'label' });
      expect(root).toMatchSnapshot();
    });
  });

  test.each([false, true])('Should render a button, case expanded %s', async expanded => {
    const { root } = await getPage({ label: 'label', expanded });
    expect(root).toMatchSnapshot();
  });

  test.each([false, true])('Should render a button, case pressed %s', async pressed => {
    const { root } = await getPage({ label: 'label', pressed });
    expect(root).toMatchSnapshot();
  });

  test('Should render a button, case controls', async () => {
    const { root } = await getPage({ label: 'label', controls: 'element-controlled' });
    expect(root).toMatchSnapshot();
  });

  test.each([true, false])('Should render a button, case haspopup %s', async haspopup => {
    const { root } = await getPage({ label: 'label', haspopup });
    expect(root).toMatchSnapshot();
  });

  test.each(buttonTypes)('Should render a button, case type %s', async type => {
    const { root } = await getPage({ label: 'label', type });
    expect(root).toMatchSnapshot();
  });

  test('Should replace classes on variant changes', async () => {
    const page = await getPage({ variant: 'primary', label: 'label' });
    const element: HTMLMgButtonElement = page.doc.querySelector('mg-button.mg-button--primary');

    expect(element).not.toBeNull();

    // Change variant
    element.variant = 'danger';
    await page.waitForChanges();

    const classPrimary = element.className.includes('.mg-button--primary');
    const classDanger = element.className.includes('.mg-button--danger');

    expect(classPrimary).toBeFalsy();
    expect(classDanger).not.toBeTruthy();
  });

  test.each(['', 'blu'])('Should throw error', async variant => {
    expect.assertions(1);
    try {
      await getPage({ variant });
    } catch (err) {
      expect(err.message).toContain('<mg-button> prop "variant" must be one of : ');
    }
  });

  test.each(['', ' ', undefined])('should throw error when using prop "isIcon" without a good prop "label"', async label => {
    expect.assertions(1);
    try {
      await getPage({ isIcon: true, label });
    } catch (err) {
      expect(err.message).toContain('<mg-button> prop "label" is mandatory when prop "isIcon" is set to true.');
    }
  });

  describe('prevent double click', () => {
    test('should NOT disable button after click', async () => {
      const page = await getPage({ identifier: 'identifier' });
      const button = page.doc.querySelector('mg-button');

      expect(page.root).toMatchSnapshot();

      button.dispatchEvent(new CustomEvent('click', { bubbles: true }));
      await page.waitForChanges();

      expect(page.root).toMatchSnapshot();

      button.disabled = true;
      await page.waitForChanges();

      expect(page.root).toMatchSnapshot();
    });

    test('should disable button after click', async () => {
      const page = await getPage({ disableOnClick: true });
      const button = page.doc.querySelector('mg-button');

      expect(page.root).toMatchSnapshot();

      button.dispatchEvent(new CustomEvent('click', { bubbles: true }));
      await page.waitForChanges();

      expect(page.root).toMatchSnapshot();

      button.disabled = false;
      await page.waitForChanges();

      expect(page.root).toMatchSnapshot();
    });

    test('should not have fn when disabled', async () => {
      const page = await getPage({
        disabled: true,
        onClick: () => {
          return false;
        },
      });

      const button = page.doc.querySelector('mg-button');

      expect(page.root).toMatchSnapshot();

      button.dispatchEvent(new CustomEvent('click', { bubbles: true }));
      await page.waitForChanges();

      const loadinClass = button.classList.contains('mg-button--loading');

      expect(loadinClass).toBeFalsy();
    });
  });

  describe.each([{ identifier: 'identifier' }, { identifier: 'identifier', disabled: true }])('keyboard', props => {
    test.each(['Space', 'Enter', 'NumpadEnter'])('Should trigger click event on keydown', async key => {
      const page = await getPage(props);
      const button = page.doc.querySelector('mg-button');

      const spy = jest.spyOn(button, 'dispatchEvent');

      button.dispatchEvent(new KeyboardEvent('keydown', { bubbles: true, key }));
      await page.waitForChanges();

      if (props.disabled) {
        expect(spy).not.lastCalledWith(expect.objectContaining({ type: 'click' }));
      } else {
        expect(spy).lastCalledWith(expect.objectContaining({ type: 'click' }));
      }
    });
    test.each(['ArrowRight', 'Tab'])('Should  NOT trigger click event on keydown', async key => {
      const page = await getPage(props);
      const button = page.doc.querySelector('mg-button');

      const spy = jest.spyOn(button, 'dispatchEvent');

      button.dispatchEvent(new KeyboardEvent('keydown', { bubbles: true, key }));
      await page.waitForChanges();

      expect(spy).not.lastCalledWith(expect.objectContaining({ type: 'click' }));
    });
  });
});
