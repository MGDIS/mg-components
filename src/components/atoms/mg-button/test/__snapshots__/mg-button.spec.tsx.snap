// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-button Should render a button, case controls 1`] = `
<mg-button>
  <!---->
  <button aria-controls="element-controlled" aria-disabled="false" aria-label="label" class="mg-button mg-button--primary" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render a button, case expanded false 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-label="label" class="mg-button mg-button--primary" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render a button, case expanded true 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-expanded="" aria-label="label" class="mg-button mg-button--primary" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render an danger button isIcon false 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-label="label" class="mg-button mg-button--danger" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render an danger button isIcon true 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-label="label" class="mg-button mg-button--danger mg-button--icon" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render an danger-alt button isIcon false 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-label="label" class="mg-button mg-button--danger-alt" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render an danger-alt button isIcon true 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-label="label" class="mg-button mg-button--danger-alt mg-button--icon" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render an flat button isIcon false 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-label="label" class="mg-button mg-button--flat" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render an flat button isIcon true 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-label="label" class="mg-button mg-button--flat mg-button--icon" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render an info button isIcon false 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-label="label" class="mg-button mg-button--info" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render an info button isIcon true 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-label="label" class="mg-button mg-button--icon mg-button--info" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render an primary button isIcon false 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-label="label" class="mg-button mg-button--primary" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render an primary button isIcon true 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-label="label" class="mg-button mg-button--icon mg-button--primary" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render an secondary button isIcon false 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-label="label" class="mg-button mg-button--secondary" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button Should render an secondary button isIcon true 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" aria-label="label" class="mg-button mg-button--icon mg-button--secondary" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button prevent double click should NOT disable button after click 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" class="mg-button mg-button--primary" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button prevent double click should NOT disable button after click 2`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" class="mg-button mg-button--primary" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button prevent double click should NOT disable button after click 3`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" class="mg-button mg-button--primary" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button prevent double click should disable button after click 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" class="mg-button mg-button--primary" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button prevent double click should disable button after click 2`] = `
<mg-button>
  <!---->
  <button aria-disabled="true" class="mg-button mg-button--loading mg-button--primary" id="identifier">
    <mg-icon icon="loader" spin=""></mg-icon>
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button prevent double click should disable button after click 3`] = `
<mg-button>
  <!---->
  <button aria-disabled="false" class="mg-button mg-button--primary" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;

exports[`mg-button prevent double click should not have fn when disabled 1`] = `
<mg-button>
  <!---->
  <button aria-disabled="true" class="mg-button mg-button--primary" id="identifier">
    <div class="mg-button__content">
      Text button
    </div>
  </button>
</mg-button>
`;
